src/Servers/Kestrel/Core/src/Internal/Http/HttpParser.cs

```diff
@@ -236,159 +234,280 @@ public unsafe bool ParseHeaders(TRequestHandler handler, ref SequenceReader<byte
+                        else
+                        {
+                            // Early memory read to hide latency
+                            var expectedCR = span[length];
+                            // Correctly has a CR, move to next
+                            length++;
+
+                            if (expectedCR != ByteCR)
                             {
-                                TakeSingleHeader(pHeader, length, handler);
+                                // Sequence needs to be CRLF not LF first.
+                                goto Reject;
                             }
-                            // Read the header successfully, skip the reader forward past the header line.
-                            reader.Advance(length);
-                            span = span.Slice(length);
+
+                            if ((uint)length < (uint)span.Length)
+                            {
+                                // Early memory read to hide latency
+                                var expectedLF = span[length];
+                                // Correctly has a LF, move to next
+                                length++;
+
+                                if (expectedLF != ByteLF)
+                                {
+                                    // Sequence needs to be CRLF and not contain an inner CR not part of terminator.
+                                    goto Reject;
+                                }
+
+                                if (length < 5)
+                                {
+                                    // Less than min possible headerSpan of 5 bytes a:b\r\n
+                                    goto Reject;
+                                }
+
+                                // Exclude the CRLF from the headerLine and parse the header name:value pair
+                                if (!TryTakeSingleHeader(handler, span[..(length - 2)]))
+                                {
+                                    // Not parsable as a valid name:value header pair.
+                                    goto Reject;
+                                }
+
+                                // Read the header successfully, skip the reader forward past the headerSpan.
+                                span = span.Slice(length);
+                                reader.Advance(length);
+
+                                goto Success;
+                            }
+                            else
+                            {
+                                // Not enough data yet.
+                                goto Continue;
+                            }
+
                         }
+
+                    // Reject and Continue are conditional jump forwards as we expect most headers to be processed first time,
+                    // so we want them to be unpredicted by an unprimied branch predictor.
+                    Reject:
+                        RejectRequestHeader(span[..length]);
```

Comment:
Do we need to add the gotos for perf? For things like `goto Reject`, it seems like it would be more straightforward to call RejectRequestHeader directly.

Reply:
One of the issues is the span slicing and then passing the sliced span to Reject; so each call to `RejectRequestHeader` adds the following chunk of asm
```asm
G_M38700_IG55:
       mov      rcx, r14
       mov      edx, r12d
       mov      eax, edx
       mov      r8d, r15d
       cmp      rax, r8
       jbe      SHORT G_M38700_IG57
						
G_M38700_IG56:
       call     System.ThrowHelper:ThrowArgumentOutOfRangeException()
       int3     
						
G_M38700_IG57:
       lea      rax, bword ptr [rsp+20H]
       mov      bword ptr [rax], rcx
       mov      dword ptr [rax+8], edx
       mov      rcx, rbx
       lea      rdx, bword ptr [rsp+20H]
       call     HttpParser:RejectRequestHeader(System.ReadOnlySpan`1[Byte]):this
       int3     
```

Reply:
Also the sliced span isn't used if `_showErrorDetails == false` which be the default state; so seems unfortunate to take the hit for uncommon header errors; for even more uncommon output of the incorrect header?

Reply:
Will try, think it would only add it twice...

Reply:
Ok, think it's doable, commit incoming

Reply:
Done